### PR TITLE
Validate instanceof types (Fixes #3033 and #762)

### LINKED ISSUE #3033: Don't allow `instanceof @Nullable C`
`instanceof` is only true for a non-null expression https://docs.oracle.com/javase/specs/jls/se13/html/jls-15.html#jls-15.20.2.

Code like this:

```java
if (nbl instanceof @Nullable C)
```

is misleading: the then-branch is not taken for a null value.

A `@Nullable` annotation on the type of an `instanceof` should raise an error.
Having a `@NonNull` annotation is redundant, and could cause a warning.

### PR DESCRIPTION
**Pull request by aditya3434**
This PR is a fix to issue#3033 (https://github.com/typetools/checker-framework/issues/3033) and issue#762 (https://github.com/typetools/checker-framework/issues/762), regarding specific type rules for the instanceof tree. In case of the Nullness Checker, the added code will issue an error if the type of the instanceof is explicitly annotated as @Nullable and will issue a warning if it is explicitly annotated as @NonNull. For other checkers, the Framework would issue a warning if the reference type is a subtype of the expression type.

Fixes #3033 and #762.

### CODE REVIEW

File: checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java [hunk 1]
```diff
@@ -380,6 +381,20 @@ public Void visitIf(IfTree node, Void p) {
         return super.visitIf(node, p);
     }
 
+    @Override
+    public Void visitInstanceOf(InstanceOfTree node, Void p) {
+        if (node.getType().getKind() == Kind.ANNOTATED_TYPE) {
+            AnnotatedTypeMirror type = atypeFactory.fromExpression((ExpressionTree) node.getType());
```
Comment (wmdietl) [inline checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java:387]: Could you:
- cast to `AnnotatedTypeTree`
- call `TreeUtils.annotationsFromTree()`
- use `AnnotationUtils.containsSame` to look for the annotations

Then you don't need to change the visibility below.

Comment (aditya3434) [reply checker/src/main/java/org/checkerframework/checker/nullness/NullnessVisitor.java]: Thanks for the help! I have used the above methods and reset the visibility of ```fromExpression``` to ```private``` in AnnotatedTypeFactory.

**Comment by aditya3434** [comment]
> Can you add a few test cases, for cases without message, cases with warnings, and cases with errors?

I added a Issue3033.java test case file that contains all the possible error and warning cases in my latest commit

File: checker/src/main/java/org/checkerframework/checker/nullness/messages.properties [hunk 2]
```diff
@@ -16,3 +16,5 @@ new.class.type.invalid=the annotations %s do not need be applied in object creat
 toArray.nullable.elements.not.newarray=call of toArray on collection of non-null elements yields an array of possibly-null elements; omit the argument to toArray or make it an explicit array constructor
 toArray.nullable.elements.mismatched.size=call of toArray on collection of non-null elements yields an array of possibly-null elements; cannot determine that the argument array has the same size as the receiver collection
 clear.system.property=call might clear a predefined system property; pass -Alint=permitClearProperty to permit it
+instanceof.nullable.error=instanceof is only true for a non-null expression
```
Comment (mernst) [inline checker/src/main/java/org/checkerframework/checker/nullness/messages.properties:19]: ".error" makes this key longer without making it more informative.  I suggest shortening the key.

File: framework/src/main/java/org/checkerframework/common/basetype/messages.properties [hunk 3]
```diff
@@ -111,3 +111,4 @@ field.invariant.not.found.superclass=the field invariant annotation is missing f
 field.invariant.not.subtype.superclass=the qualifier for field %s is not a subtype of the qualifier in the superclass field invariant%nfound: %s%nsuperclass type: %s
 
 invalid.annotation.location.bytecode=found annotation in unexpected location in bytecode on element: %s %nUse -AignoreInvalidAnnotationLocations to suppress this warning
+operand.instanceof.subtype=reference type is a subtype of the operand type.%nOperand Type : %s%nReference Type : %s
```
Comment (mernst) [inline framework/src/main/java/org/checkerframework/common/basetype/messages.properties:114]: I find this error message confusing.  The only point of performing `x instanceof C` is when `C` is a subtype of `x`'s compile-time type, so it's weird for the message to say only that.

I think this message is trying to say that, at run time, `instanceof` will not check the annotations.  Please make it do so.  That is similar to the "cast.unsafe" message earlier in the file.  You could make the error key "instanceof.unsafe" for parallelism.

Also, please use standard terms in the diagnostic message.  The JLS uses "expression" where you use "operand"

**Review by mernst — APPROVED** [review]
Thanks for this improvement!  It's great to have those issues fixed.

By the way, thanks for addressing my comments.  I had not noticed that you did so.  As a general rule, when you have addressed all comemnts, please ping for another round of review.

**Pull request merged**